File: examples/quickstart/typescript/03_signed_messages.ts
#!/usr/bin/env npx tsx
/**
 * Quickstart 03: HMAC-SHA256 and Ed25519 signing with tamper detection.
 *
 * Run: npx tsx 03_signed_messages.ts
 */
import {
  createIdentity,
  createMessage,
  signMessage,
  verifyMessage,
  generateEd25519KeyPair,
  signEd25519,
  verifyEd25519,
  MessageIntent,
} from "@roar-protocol/sdk";

// --- HMAC-SHA256 signing (symmetric, shared secret) ---
const alice = createIdentity("alice", { capabilities: ["crypto"] });
const bob = createIdentity("bob", { capabilities: ["crypto"] });
const secret = process.env.ROAR_SIGNING_SECRET || "quickstart-demo-key";

const msg = signMessage(
  createMessage(alice, bob, MessageIntent.NOTIFY, {
    text: "Hello Bob, this message is tamper-proof!",
  }),
  secret,
);

console.log(`HMAC signature: ${msg.auth.signature?.substring(0, 50)}...`);
console.log(`HMAC-SHA256: verified = ${verifyMessage(msg, secret)}`);

// Tamper detection
msg.payload.text = "TAMPERED!";
console.log(`HMAC-SHA256: tamper detected = ${!verifyMessage(msg, secret)}`);

// --- Ed25519 signing (asymmetric, key pair) ---
const keyPair = generateEd25519KeyPair();
console.log(`\nEd25519 public key: ${keyPair.publicKey.substring(0, 32)}...`);

const agent = createIdentity("ed25519-agent", { publicKey: keyPair.publicKey });
const peer = createIdentity("peer");

const msg2 = createMessage(agent, peer, MessageIntent.DELEGATE, {
  task: "verify-me",
});
const signed = signEd25519(msg2, keyPair.privateKey);
console.log(`Ed25519 signature: ${signed.auth.signature?.substring(0, 50)}...`);
console.log(`Ed25519: verified = ${verifyEd25519(signed)}`);

console.log("\nAll signing demos passed!");

File: examples/quickstart/python/03_signed_messages.py
#!/usr/bin/env python3
"""Quickstart 03: HMAC-SHA256 and Ed25519 signing and verification.

Run: python 03_signed_messages.py
"""
import os
from roar_sdk import AgentIdentity, MessageIntent, ROARMessage
from roar_sdk.signing import generate_keypair, sign_ed25519, verify_ed25519

# --- HMAC-SHA256 signing (symmetric, shared secret) ---
alice = AgentIdentity(display_name="alice", capabilities=["crypto"])
bob = AgentIdentity(display_name="bob", capabilities=["crypto"])
secret = os.environ.get("ROAR_SIGNING_SECRET", "quickstart-demo-key")

msg = ROARMessage(
    **{"from": alice, "to": bob},
    intent=MessageIntent.NOTIFY,
    payload={"text": "Hello Bob, this message is tamper-proof!"},
)

msg.sign(secret)
print(f"HMAC signature: {msg.auth['signature'][:50]}...")
assert msg.verify(secret), "HMAC verification failed!"
print("HMAC-SHA256: verified OK")

# Tamper detection
msg.payload["text"] = "TAMPERED!"
assert not msg.verify(secret), "Should have detected tampering!"
print("HMAC-SHA256: tamper detected OK")

# --- Ed25519 signing (asymmetric, key pair) ---
priv_key, pub_key = generate_keypair()
print(f"\nEd25519 public key: {pub_key[:32]}...")

agent = AgentIdentity(display_name="ed25519-agent", public_key=pub_key)
peer = AgentIdentity(display_name="peer")

msg2 = ROARMessage(
    **{"from": agent, "to": peer},
    intent=MessageIntent.DELEGATE,
    payload={"task": "verify-me"},
)

sign_ed25519(msg2, priv_key)
print(f"Ed25519 signature: {msg2.auth['signature'][:50]}...")
assert verify_ed25519(msg2), "Ed25519 verification failed!"
print("Ed25519: verified OK")

print("\nAll signing demos passed!")
